From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: "Jan Alexander Steffens (heftig)" <heftig@archlinux.org>
Date: Sat, 4 Jan 2025 06:52:13 +0100
Subject: [PATCH] compiler: Use target-specific GCC linkers

Otherwise it defaults to 'cc', which is inappropriate when we're
cross-compiling.
---
 .../rustc_target/src/spec/targets/aarch64_unknown_linux_gnu.rs   | 1 +
 .../rustc_target/src/spec/targets/aarch64_unknown_linux_musl.rs  | 1 +
 compiler/rustc_target/src/spec/targets/i686_unknown_linux_gnu.rs | 1 +
 .../rustc_target/src/spec/targets/x86_64_unknown_linux_gnu.rs    | 1 +
 .../rustc_target/src/spec/targets/x86_64_unknown_linux_musl.rs   | 1 +
 5 files changed, 5 insertions(+)

diff --git a/compiler/rustc_target/src/spec/targets/aarch64_unknown_linux_gnu.rs b/compiler/rustc_target/src/spec/targets/aarch64_unknown_linux_gnu.rs
index a4a5ee6f2c9c..434e70796a63 100644
--- a/compiler/rustc_target/src/spec/targets/aarch64_unknown_linux_gnu.rs
+++ b/compiler/rustc_target/src/spec/targets/aarch64_unknown_linux_gnu.rs
@@ -39,6 +39,7 @@ pub(crate) fn target() -> Target {
                 | SanitizerSet::HWADDRESS
                 | SanitizerSet::REALTIME,
             supports_xray: true,
+            linker: Some("aarch64-linux-gnu-gcc".into()),
             ..base::linux_gnu::opts()
         },
     }
diff --git a/compiler/rustc_target/src/spec/targets/aarch64_unknown_linux_musl.rs b/compiler/rustc_target/src/spec/targets/aarch64_unknown_linux_musl.rs
index 93319d23bf07..3849e381566c 100644
--- a/compiler/rustc_target/src/spec/targets/aarch64_unknown_linux_musl.rs
+++ b/compiler/rustc_target/src/spec/targets/aarch64_unknown_linux_musl.rs
@@ -14,6 +14,7 @@ pub(crate) fn target() -> Target {
         | SanitizerSet::LEAK
         | SanitizerSet::MEMORY
         | SanitizerSet::THREAD;
+    base.linker = Some("aarch64-linux-gnu-gcc".into());
 
     // FIXME(compiler-team#422): musl targets should be dynamically linked by default.
     base.crt_static_default = true;
diff --git a/compiler/rustc_target/src/spec/targets/i686_unknown_linux_gnu.rs b/compiler/rustc_target/src/spec/targets/i686_unknown_linux_gnu.rs
index 77e77338d0ed..4a2f1a739fb8 100644
--- a/compiler/rustc_target/src/spec/targets/i686_unknown_linux_gnu.rs
+++ b/compiler/rustc_target/src/spec/targets/i686_unknown_linux_gnu.rs
@@ -23,6 +23,7 @@ pub(crate) fn target() -> Target {
     base.add_pre_link_args(LinkerFlavor::Gnu(Cc::Yes, Lld::No), &["-m32"]);
     base.stack_probes = StackProbeType::Inline;
     base.supports_fentry = true;
+    base.linker = Some("x86_64-linux-gnu-gcc".into());
 
     Target {
         llvm_target: "i686-pc-linux-gnu".into(),
diff --git a/compiler/rustc_target/src/spec/targets/x86_64_unknown_linux_gnu.rs b/compiler/rustc_target/src/spec/targets/x86_64_unknown_linux_gnu.rs
index 36f41231ae00..078c614b0d4b 100644
--- a/compiler/rustc_target/src/spec/targets/x86_64_unknown_linux_gnu.rs
+++ b/compiler/rustc_target/src/spec/targets/x86_64_unknown_linux_gnu.rs
@@ -20,6 +20,7 @@ pub(crate) fn target() -> Target {
         | SanitizerSet::REALTIME;
     base.supports_fentry = true;
     base.supports_xray = true;
+    base.linker = Some("x86_64-linux-gnu-gcc".into());
 
     Target {
         llvm_target: "x86_64-pc-linux-gnu".into(),
diff --git a/compiler/rustc_target/src/spec/targets/x86_64_unknown_linux_musl.rs b/compiler/rustc_target/src/spec/targets/x86_64_unknown_linux_musl.rs
index a664f2247991..2e9007ba05e4 100644
--- a/compiler/rustc_target/src/spec/targets/x86_64_unknown_linux_musl.rs
+++ b/compiler/rustc_target/src/spec/targets/x86_64_unknown_linux_musl.rs
@@ -19,6 +19,7 @@ pub(crate) fn target() -> Target {
     base.supports_xray = true;
     // FIXME(compiler-team#422): musl targets should be dynamically linked by default.
     base.crt_static_default = true;
+    base.linker = Some("x86_64-linux-gnu-gcc".into());
 
     Target {
         llvm_target: "x86_64-unknown-linux-musl".into(),
